ql/src/java/org/apache/hadoop/hive/ql/optimizer/calcite/rules/HiveAggregateJoinTransposeRule.java

```diff
@@ -303,6 +305,90 @@ public void onMatch(RelOptRuleCall call) {
     }
   }
 
+  /**
+   * Determines weather the give grouping is unique.
+   *
+   * Consider a join which might produce non-unique rows; but later the results are aggregated again.
+   * This method determines if there are sufficient columns in the grouping which have been present previously as unique column(s).
+   */
+  private boolean isGroupingUnique(RelNode input, ImmutableBitSet groups) {
+    if (groups.isEmpty()) {
+      return false;
+    }
+    RelMetadataQuery mq = input.getCluster().getMetadataQuery();
+    Set<ImmutableBitSet> uKeys = mq.getUniqueKeys(input);
+    for (ImmutableBitSet u : uKeys) {
+      if (groups.contains(u)) {
+        return true;
+      }
+    }
+    if (input instanceof Join) {
+      Join join = (Join) input;
+      RexBuilder rexBuilder = input.getCluster().getRexBuilder();
+      SimpleConditionInfo cond = new SimpleConditionInfo(join.getCondition(), rexBuilder);
+
+      if (cond.valid) {
+        ImmutableBitSet newGroup = groups.intersect(ImmutableBitSet.fromBitSet(cond.fields));
+        RelNode l = join.getLeft();
+        RelNode r = join.getRight();
+
+        int joinFieldCount = join.getRowType().getFieldCount();
+        int lFieldCount = l.getRowType().getFieldCount();
+
+        ImmutableBitSet groupL = newGroup.get(0, lFieldCount);
+        ImmutableBitSet groupR = newGroup.get(lFieldCount, joinFieldCount).shift(-lFieldCount);
+
+        if (isGroupingUnique(l, groupL)) {
```

Comment:
this method does a bit different thing - honestly I feeled like I'm in trouble when I've given this name to it :)

this method checks if the given columns contain an unique column somewhere in the covered joins; (this still sound fuzzy) so let's take an example

consider:
```
select c_id, sum(i_prize) from customer c join item i on(i.c_id=c.c_id)
```

* do an aggregate grouping by the column C_ID  ; and sum up something 
* below is a join which joins by C_ID
* asking wether C_ID  is a unique column on top of the join is false; but there is subtree in which C_ID is unique => so if we push the aggregate on that branch the aggregation will be a no-op

I think this case is not handled by `areColumnsUnique`

Reply:
Could you execute `areColumnsUnique` on the join input then? Wouldn't that simplify this logic?

Reply:
this method recursively checks that the above condition is satisfied or not - that's why it needs to call itself

Reply:
As you go down recursively, you may start finding `HepRelVertex` as the rel node. I think you hit the `instanceof Project` below only for the first project because you create it using the builder.
That is why I think once you have gone through the first join, you will not hit another join; you could simply call `areColumnsUnique` in these `if` clauses, which could potentially uncover new cases.

I may be wrong though, I just wanted to leave you a final note to make sure it was clear what I meant.

Reply:
That could be done; and I'm sure it was true in this - but this logic will work better if it could walk down as many joins as it could - we might have an aggregate on top in the meantime a bunch of joins under it...so I feel that it will be beneficial to retain it.
I feeled tempted to write a RelMd handler - however I don't think I could just introduce a new one easily.
RelShuttle doesn't look like a good match - I'll leave it as a set of `instanceof` calls for now.

I'll upload a new patch to see if digging deeper in the tree could do more or not.

Reply:
OK. If it turns out there are many changes and it may need some time to be fixed, feel free to defer to follow-up JIRA and let's merge this one.